File: leetcode/3212-CountSubmatricesWithEqualFrequencyOfXAndY/solution.go
// Compute complexity = O(N*M)
// Space complexity = O(N*M)

type Pair struct {
    First int
    Second int
}

func (self *Pair) Update(ch byte) {
    if ch == 'X' {
        self.First++
    } else if (ch == 'Y') {
        self.Second++
    }
}

func (self *Pair) Add(val *Pair) {
    self.First += val.First
    self.Second += val.Second
}

func numberOfSubmatrices(grid [][]byte) int {
    // get dimensions
    N := len(grid)
    if 0 == N {
        return 0
    }
    M := len(grid[0])
    if  0 == M {
        return 0
    }

    dic := make([][]Pair, N)
    for k, _ := range dic {
        dic[k] = make([]Pair, M)
    }

    // populate the first column of each row according to grid
    for i := 0; i < N; i++ {
        dic[i][0].Update(grid[i][0])
    }
    // then calculate the raw accumulation of occurances
    for i := 0; i < N; i++ {
        for j := 1; j < M; j++ {
            dic[i][j] = dic[i][j-1]
            dic[i][j].Update(grid[i][j])
        }
    }

    // then calculate the column accumulation of occurances
    for i := 1; i < N; i++ {
        for j := 0; j < M; j++ {
            dic[i][j].Add(&dic[i-1][j])
        }
    }

    // now iterate over the accumulation of X & Y values to calculate the
    // number of valid submatrices
    noOfSub := 0
    for _, raw := range dic {
        for _, v := range raw {
            if v.First == v.Second && v.First > 0 {
                noOfSub++
            }
        }
    }


    return noOfSub
}

File: leetcode/3212-CountSubmatricesWithEqualFrequencyOfXAndY/solution.cpp
// Compute complexity = O(N*M)
// Space complexity = O(N*M)

class Solution {
public:
    int numberOfSubmatrices(vector<vector<char>>& grid) {
        // get dimensions and do a asanity check
        const auto N = grid.size();
        if (0 == N) {
            return 0;
        }
        const auto M = grid.front().size();
        if (0 == M) {
            return 0;
        }

        // create a balance matrix that will contain pairs
        // where it will hold the cumulative 'X' and 'Y' occurances in the corresponding row
        // with adding the values with previous column
        vector<vector<pair<int, int>>> balance(N, vector<pair<int, int>>(M, {0, 0}));
        for (int i = 0; i < N; ++i){
            const auto ch = grid[i][0];
            balance[i][0].first =  (ch == 'X' ? 1 : 0);
            balance[i][0].second = (ch == 'Y' ? 1 : 0);    
        }
        for (int i = 0; i < N; ++i) {
            for (int j = 1; j < M; ++j) {
                const auto ch = grid[i][j];
                balance[i][j].first = balance[i][j-1].first + (ch == 'X' ? 1 : 0);
                balance[i][j].second = balance[i][j-1].second + (ch == 'Y' ? 1 : 0);
            }
        }

        // find overall cumulatives with adding the previous raw
        for (int i = 1; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                balance[i][j].first += balance[i-1][j].first;
                balance[i][j].second += balance[i-1][j].second;
            }
        }

        // calculate number of solutions where finding the accumulated X & Ys of
        // previous columns. Check also occurance of X as well
        int noOfSol = 0;
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                if (balance[i][j].first == balance[i][j].second && balance[i][j].first > 0) {
                    noOfSol++;
                }
            }
        }

        return noOfSol;
    }
};